httomo/ui_layer.py

```diff
@@ -0,0 +1,112 @@
+        self.repo = MethodDatabaseRepository()
+        self.tasks_file_path = tasks_file_path
+        self.in_data_file = in_data_file_path
+        self.comm = comm
+        
+        root, ext = os.path.splitext(self.tasks_file_path)
+        if ext in ['.yaml', '.yaml'.upper()]:
+            # loading yaml file with tasks provided
+            self.PipelineStageConfig = _yaml_loader(self.tasks_file_path)[0]
+        elif ext in ['.py', '.py'.upper()]:
+            # loading python file with tasks provided
+            self.PipelineStageConfig = _python_tasks_loader(self.tasks_file_path)
+        else:
+            raise ValueError(
+                f"The extension {ext} of the file {root} with tasks is unknown."
+            )
+
+    def build_pipeline(self) -> Pipeline:
+        side_outputs_collect: list = [] # saves [task_no, id, side_outputs] for tasks with side_outputs
+        save_result_collect: list = [] 
+        methods_list: list = []
+        for task_no, task_conf in enumerate(self.PipelineStageConfig):
+            if "loaders" in task_conf['module_path']:
+                task_conf['parameters']['in_file'] = self.in_data_file
+                # unpack params and initiate a loader
+                loader =  make_loader(
+                        self.repo,
+                        task_conf['module_path'],
+                        task_conf['method'],
+                        self.comm,
+                        **task_conf['parameters'],
+                        )
+            else:
+                if "parameters" not in task_conf:
+                    task_conf['parameters'] = {}
+                if "save_result" not in task_conf:
+                    save_result_collect.append(False)
+                else:
+                    save_result_collect.append(task_conf['save_result'])
+                if "side_outputs" not in task_conf:
+                    task_conf['side_outputs'] = {}
+                else:
+                    side_outputs_collect.append([task_no, task_conf['id'], task_conf['side_outputs']])
+                # check if there is a reference to side_outputs to cross-link
+                for key, value in task_conf['parameters'].items():
+                    if isinstance(value, str) and value is not None:
+                        if value.find('${{') != -1:                            
+                            result_extr = re.search(r"\{([A-Za-z0-9_.]+)\}", value)
+                            internal_expression = result_extr.group(1)
+                            (ref_id, side_str, ref_arg) = internal_expression.split(".")
+                            # lets find the referred id in "side_outputs_collect"
+                            for items in side_outputs_collect:
+                                if items[1] == ref_id:
+                                    # refer to methods_list[items[0]-1]
+                                    task_conf['parameters'][key] = OutputRef(methods_list[items[0]-1], ref_arg)
```

Comment:
Maybe not now, but at some point, it may be worth refactoring this to make it easier to read/understand.

Since this looks related to checking the validity of references to side outputs, extracting the logic to a separate function during refactoring may also prove useful for the discussion point we had the other day about wanting the YAML checker to check the validity of references in a pipeline YAML file?

Reply:
I agree with this. I was looking at this same part yesterday, and thinking of refactoring to see if that could be tested in a more isolated way.

Reply:
yes, good point. Feel free to move to a separate function that assigns the side_outputs references. @jessicavers I guess you could do that in your branch for yaml checker?